.github/workflows/hrm-lambda-release.yml

```diff
@@ -0,0 +1,113 @@
+
+      send-slack-message:
+        description: 'Specifies if should send a Slack message at the end of successful run. Defaults to true'
+        required: true
+        type: boolean
+        default: true
+
+  workflow_call:
+    secrets:
+      AWS_ACCESS_KEY_ID:
+        required: true
+      AWS_SECRET_ACCESS_KEY:
+        required: true
+      # Should probably use the passed in region for everything but the GITHUB_ACTIONS_SLACK_BOT_TOKEN SSM parameter only exists in us-east-1
+      AWS_DEFAULT_REGION:
+        required: true
+
+env:
+  GITHUB_SHA: ${{ github.sha }}
+
+jobs:
+  build-lambda:
+    name: Build And Deploy Lambda
+    runs-on: ubuntu-latest
+
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v2
+
+      - name: Setup Node.js
+        uses: actions/setup-node@v3
+        with:
+          node-version: 14.x
+
+      #TODO: Move to a matrix based region split for the ECR/update lambda steps
+      - name: Set Env Based on Environment
+        # this is overcomplicated to allow for testing with a push event
+        run: |
+          INPUT_ENVIRONMENT=${{ inputs.environment }}
+          INPUT_LAMBDA_NAME=${{ inputs.lambda_name }}
+          ENVIRONMENT=${INPUT_ENVIRONMENT:-development}
+          LAMBDA_NAME=${INPUT_LAMBDA_NAME:-retrieve-transcript}
+
+          echo ENVIRONMENT=$ENVIRONMENT >> $GITHUB_ENV
+          echo LAMBDA_NAME=$LAMBDA_NAME >> $GITHUB_ENV
+
+          case "$ENVIRONMENT" in development) region="us-east-1" ;; staging) region="eu-west-1" ;; production) region="eu-west-1" ;; *) echo "Invalid environment: ${{ inputs.environment }}" && exit 1 ;; esac
+          echo AWS_REGION=$region >> $GITHUB_ENV
+          echo AWS_DEFAULT_REGION=$region >> $GITHUB_ENV
+          echo DOCKER_IMAGE_BASE=712893914485.dkr.ecr.${region}.amazonaws.com/${ENVIRONMENT}/hrm-lambdas- >> $GITHUB_ENV
+
+      - name: Compile Lambda
+        run: |
+          cd lambdas
+          npm ci
+          npm run compile:${{ env.LAMBDA_NAME }}
+
+      - name: Configure AWS credentials
+        uses: aws-actions/configure-aws-credentials@v1
+        with:
+          aws-access-key-id: ${{ secrets.AWS_ACCESS_KEY_ID }}
+          aws-secret-access-key: ${{ secrets.AWS_SECRET_ACCESS_KEY }}
+          aws-region: ${{ env.AWS_REGION }}
+
+      - name: Login to Amazon ECR
+        id: login-ecr
+        uses: aws-actions/amazon-ecr-login@v1
+
+      # build and push the image
+      - name: Build and Push Docker Image
+        uses: docker/build-push-action@v2
+        with:
+          context: ./lambdas/src/${{ env.LAMBDA_NAME }}
+          file: ./lambdas/Dockerfile
+          push: true
+          tags: ${{ env.DOCKER_IMAGE_BASE }}${{ env.LAMBDA_NAME }}:live
+
+      - name: Update Lambda and Publish
+        run: |
+          aws lambda update-function-code --function-name ${{ env.ENVIRONMENT }}-hrm-contact-job-${{ env.LAMBDA_NAME }} --image-uri ${{ env.DOCKER_IMAGE_BASE }}${{ env.LAMBDA_NAME }}:live --publish
```

Comment:
If this assumes that the function to be deployed will be prefixed with the `${{ env.ENVIRONMENT }}-hrm-contact-job-`, I would prefer to rename this workflow to something like "Deploy hrm-contact-job Lambda". The other option could be remove that hard coded chunk and rely on the lambdas source being called after `hrm-contact-job-xxxxxx`.

Reply:
Namespacing the lambda function allows for simplified IAM permissions like:

```
{
   "Version": "2012-10-17",
   "Statement": [
       {
           "Sid": "VisualEditor0",
           "Effect": "Allow",
           "Action": [
               "lambda:UpdateFunctionCode"
           ],
           "Resource": [
               "arn:aws:lambda:us-east-1:712893914485:function:development-*",
               "arn:aws:lambda:us-east-1:712893914485:function:staging-*",
               "arn:aws:lambda:us-east-1:712893914485:function:production-*"
           ]
       }
   ]
}
```

I'd strongly suggest we move towards this style of aws naming. For ARNS that support `/`, I would generally recommend names like `development/some-name` so you can apply env based IAM like;

```
"arn:aws:ecr:us-east-1:712893914485:repository/development/*",
```

Reply:
You can get more granular with the permissions if someone (or some service) only need access to hrm functions in development you can just give them access to: `arn:aws:lambda:us-east-1:712893914485:function:development-hrm-*` if they/it just needs access to contact jobs in hrm in development you can do `arn:aws:lambda:us-east-1:712893914485:function:development-hrm-contact-jobs-*`.

Using naming conventions to extend IAM permissions reduces the need to modify a policy every time a developer adds something new to the stack... which gets VERY cumbersome as infrastructures grow more complex. The order of the naming becomes very important when you start thinking this way about permissions.